src/modules/claims/claims.service.ts

```diff
@@ -372,46 +413,96 @@ export class ClaimsService {
     return v4();
   }
 
+  validatePublishPublicClaimRequest(
+    registrationTypes: RegistrationTypes[],
+    token?: string,
+    claimType?: string
+  ) {
+    if (registrationTypes.includes(RegistrationTypes.OnChain) && !claimType) {
+      throw new Error('claimType required for on-chain registration');
+    }
+    if (registrationTypes.includes(RegistrationTypes.OffChain) && !token) {
+      throw new Error('Token required for on-chain registration');
+    }
+  }
+
   /**
    * publishPublicClaim
    *
    * @description store claim data in ipfs and save url to DID document services
    * @returns ulr to ipfs
    *
    */
-  async publishPublicClaim({ token }: { token: string }) {
-    const payload = (await this._didRegistry.decodeJWTToken({ token })) as {
-      iss: string;
-      sub: string;
-      claimData: ClaimData;
-    };
-    const { iss, claimData } = payload;
-    let sub = payload.sub;
-    // Initialy subject was ignored because it was requester
-    if (!sub || sub.length === 0 || !isValidDID(sub)) {
-      sub = this._signerService.did;
-    }
-    const verifiedDid = await this._didRegistry.verifyPublicClaim(token, iss);
-    if (!verifiedDid || !compareDID(verifiedDid, iss)) {
-      throw new Error('Incorrect signature');
+  async publishPublicClaim({
+    token,
+    claimType,
+    registrationTypes = [RegistrationTypes.OffChain],
+  }: {
+    token?: string;
+    claimType?: string;
+    registrationTypes?: RegistrationTypes[];
+  }) {
+    this.validatePublishPublicClaimRequest(registrationTypes, token, claimType);
+    let url: string | undefined = undefined;
+    if (registrationTypes.includes(RegistrationTypes.OnChain)) {
+      const claims = await this.getClaimsBySubject({
```

Comment:
This would be more structured if `if registrationTypes.includes(RegistrationTypes.OnChain) {` was followed by `registerOnChain({claim})`. Issuance of claim can be checked in `validatePublishPublicClaimRequest`

Reply:
+1